### DISCUSSION TITLE
Stream upload problem for binary files without newline

### DISCUSSION THREAD

**Discussion by ysn2233** [encode/httpx#2504] [Potential Issue] [answered]
``` python
client = httpx.Client()
client.put(url=url, data=open("filename", "rb"))
```
When I use httpx to upload a large binary file(>2G), it will lead to **OverflowError: string longer than 2147483647 bytes**. It is because my binary file don't contain '\n' byte and httpx stream the data with default iterator of the fileobj(`_io.BufferedReader`). It will read chunk by '\n', so in my case all bytes will be read into memory and it is definetely not reasonable. This problem may also occur in case of a file with an extremely long line.

I can externally solve this problem by  wrapper a custom class extended `BufferedReader` and rewrite `__next__`. However do you think we should support fixed-size chunk streaming upload in httpx? My solution is to rewrite `__iter__` in class `IteratorByteStream` like that:
``` python
    def __iter__(self) -> Iterator[bytes]:
        if self._is_stream_consumed and self._is_generator:
            raise StreamConsumed()

        self._is_stream_consumed = True

        # old
        # for part in self._stream:
            # yield part
        # new
        while True:
            part = self._stream.read(self.block_size)
            if not part:
                break
            yield part
```
Or some better design could be made.

**Comment by lovelydinosaur** [comment, marked as answer]
* Double check you're using the latest version of `httpx`.
* You want `content=...` for byte uploads not `data=...`.

If you take a look at the internals you'll see that we *do* prefer `.read(chunk_size)` where possible...

https://github.com/encode/httpx/blob/b82fbe2c139162e10d1a3ee1b4579cc8d72b3535/httpx/_content.py#L50-L64

We didn't get this right in the past, but it was fixed in https://github.com/encode/httpx/pull/1948
